mooncake-store/include/allocation_strategy.h

```diff
@@ -80,102 +184,96 @@ class RandomAllocationStrategy : public AllocationStrategy {
+                used_segments.contains(preferred_segment)) {
+                // Skip excluded and used segments
+                continue;
+            }
+
+            auto buffer = allocateSingle(allocator_manager, preferred_segment,
+                                         slice_length, generator);
+            if (buffer) {
+                replicas.emplace_back(std::move(buffer),
+                                      ReplicaStatus::PROCESSING);
+                if (replicas.size() == replica_num) {
+                    return replicas;
+                }
+
+                // Add preferred segment to used_segments on allocation success
+                used_segments.insert(preferred_segment);
             }
-            return replicas;
         }
 
-        static thread_local std::mt19937 generator(clock());
-        std::uniform_int_distribution<size_t> distribution(
-            0, allocators_by_name.size() - 1);
+        // If replica_num is not satisfied, allocate the remaining replicas
+        // randomly.
+        std::uniform_int_distribution<size_t> distribution(0, names.size() - 1);
         size_t start_idx = distribution(generator);
 
-        // Get iterator to the starting point
-        auto start_it = allocators_by_name.begin();
-        std::advance(start_it, start_idx);
-
-        auto it = start_it;
-        size_t max_retry = std::min(kMaxRetryLimit, allocators_by_name.size());
-        size_t retry_count = 0;
-
-        // Try to allocate remaining replicas, starting from random position
-        // TODO: Change the segment data structure to avoid traversing the
-        // entire map every time
-        while (replicas.size() < config.replica_num &&
-               retry_count < max_retry) {
-            // Skip preferred segment if it was already allocated
-            if (it->first != config.preferred_segment) {
-                // Try each allocator in this segment
-                bool allocated = false;
-                for (auto& allocator : it->second) {
-                    if (auto buffer = allocator->allocate(slice_length)) {
-                        replicas.emplace_back(std::move(buffer),
-                                              ReplicaStatus::PROCESSING);
-                        // Allocate at most one replica per segment
-                        allocated = true;
-                        break;
-                    }
-                }
-                if (!allocated) {
-                    ++retry_count;
-                }
-            }
-            // Move to next segment (circular)
-            ++it;
-            if (it == allocators_by_name.end()) {
-                it = allocators_by_name.begin();
+        const size_t max_retry = std::min(kMaxRetryLimit, names.size());
+        size_t try_count = 0;
+
+        while (replicas.size() < replica_num && try_count < max_retry) {
+            auto index = start_idx % names.size();
+            start_idx++;
+            try_count++;
```

Comment:
Nit: Would it be better to increase `try_count` only when *not* skipping excluded and used segments? The meaning of `try_count` here is the number of attempts to make after excluding the preferred and excluded segments.

Reply:
`try_count` serves two purposes here, for one, it ensures the number of allocation attempts not exceed `max_retry`, for the other, it ensures that we only try allocation once for each segment. With that saying, maybe we should use two different counters for the two purposes?